README.md
[![Travis CI](https://img.shields.io/travis/skx/pam_pwnd/master.svg?style=flat-square)](https://travis-ci.org/skx/pam_pwnd)
[![license](https://img.shields.io/github/license/skx/pam_pwnd.svg)](https://github.com/skx/pam_pwnd/blob/master/LICENSE)
[![Release](https://img.shields.io/github/release/skx/pam_pwnd.svg)](https://github.com/skx/pam_pwnd/releases/latest)

# pam_pwnd

This repository contains a simple PAM module for testing whether a
password being used for authentication has been listed in the
[have I been pwned](https://haveibeenpwned.com/) database.

Note that in the documentation here we focus upon ensuring that a password used for `sudo` has not been compromised, but PAM-modules can be used for many purposes, from handling SSH-access, to permitting HTTP-based authentication.  There is nothing `sudo`-specific about our code so this module can be useful in many contexts.


## Sponsorship

The development of this module was sponsored by three individuals who made charitable donations.  (Anonymous primarily because I didn't ask for permission to name them publicly.)

If you wish to "sponsor this" software, and be listed here, just [email me](https://steve.kemp.fi/) a receipt of your donation.  I support the [RNLI](https://en.wikipedia.org/wiki/Royal_National_Lifeboat_Institution), but feel free to pick whatever charity you wish.

The code is released under the [BSD-license](LICENSE) so you can fork it, improve it, use it, and enjoy it!  Feel free to report bugs, or feature-suggestions on the [issue-page](https://github.com/skx/pam_pwnd/issues).



## Compilation

These are the dependencies I expect you would need for compiling the project:

* For fetching a remote URI we use `libcurl`:
  * `apt-get install libcurl4-gnutls-dev`
* For compiling PAM modules you'll need the appropriate development package:
  * `apt-get install libpam0g-dev`

Assuming you have the dependencies installed then compilation should only require a simple `make`:

    $ make
    gcc -fPIC -c pam_pwnd.c -lpam -lpam_misc -lpamc
    gcc -fPIC -c pwn_chk.c  -lcurl
    gcc -fPIC -c sha1.c
    ld -x --shared -o pam_pwnd.so pam_pwnd.o pwn_chk.o sha1.o -lcurl -lpam -lpam_misc -lpamc

For completeness you can also run the basic test-cases included in the repository, but note that to do that you will require network-access:

    $ make test
    ./pam_test
    ..

(This might be an issue if you run the tests as part of a build-process upon a CI/CD system which doesn't permit outgoing network access.)



## Installation & Configuration


Once you have compiled the code you should copy the resulting file `pam_pwnd.so` to the appropriate PAM-directory upon your system.  In my case that means running this command:

    sudo install pam_pwnd.so  /lib/x86_64-linux-gnu/security/

The final step is to enable the module, by editing the appropriate PAM configuration file.

In my case I'm using SSH keys for authentication, and I'm only concerned with ensuring that no known-bad passwords are used with `sudo`.  I append the following line to `/etc/pamd.d/sudo`:

    auth   required   pam_pwnd.so  try_first_pass

The complete file, on an Ubuntu system, might then look like this:

      #%PAM-1.0
      session    required   pam_env.so readenv=1 user_readenv=0
      session    required   pam_env.so readenv=1 envfile=/etc/default/locale user_readenv=0
      @include common-auth
      @include common-account
      @include common-session-noninteractive
      auth   required   pam_pwnd.so  try_first_pass

Upon the "stretch" release of Debian GNU/Linux the file has these contents:

      #%PAM-1.0

      @include common-auth
      @include common-account
      @include common-session-noninteractive
      auth   required   pam_pwnd.so  try_first_pass

Regardless of what your file looks like, once you've added the reference to `pam_pwnd.so`, you should then be ready to test the module hasn't broken your system by reseting the `sudo` cache, and re-authenticating:

     frodo ~ $ sudo -k
     frodo ~ $ sudo su -
     [sudo] password for skx:
     root@frodo:~#

Assuming nothing is broken you should:

* Be prompted for your password.
  * Only once.
* Receive your root-prompt.
* See the results of the module logged to syslog.

If things are horribly broken, such that you get segfaults or failures from _this_ module then you will probably be unable to run `sudo` to fix them, so for the duration of any installation you should ensure you have an open terminal/connection with `root` privileges.

The module will log its results to syslog, search for `pam_pwnd` to see them.



## Security Notes

The code makes a single outgoing HTTP-request for each authentication
request:

* The outgoing request contains the first five characters of your __hashed__ password.
   * i.e. If you password is "secret" it is first hashed to `e5e9fa1ba31ecd1ae84f75caaa474f3a663f05f4`.
   * Then an outgoing request is made with the characters `e5e9f`.

If the API-lookup request fails then we default to failing-open, allowing the authentication to proceed.   (We assume other modules will actually validate the password, if we allowed a failure to invoke the API we'd deny all PAM-based operations in the event your DNS, networking, or similar things were broken.)

There are zero memory allocations in this module, which should ensure that we don't leak anything.  Instead we generate a single temporary file to hold the results of our HTTP-response, and that temporary file is cleaned up after use.


## Testing Notes

There is a simple test-driver included in this project which exercises some of
the code, it is not designed to be a complete test-case, nor to perform exhaustive testing.

If you're planning to submit pull-requests that change the code you should ensure the tests pass even with your additions:

    $ make test


## Feedback

Bug reports welcome.


Steve
--


debian/changelog
libpam-pwnd (0.1) unstable; urgency=low

  * Initial package.

 -- Steve Kemp <steve@steve.org.uk>  Mon, 17 Sep 2018 19:19:19 +0200


pam_test.c
/*
 * pam_test.c : Simple test-cases to exercise our code at least a little.
 */


#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>

#include "sha1.h"

/*
 * The lookup of the password is handled via this external function.
 */
extern int was_leaked(char *hash);


/*
 * This structure describes a simple SHA1-test.
 *
 * We provide an input and ensure that the output hash matches
 * the given output.
 */
typedef struct sha1_test_case
{
    /*
     * Input to the test-case.
     */
    char *input;

    /*
     * Expected output.
     */
    char *output;

} sha1_test_case;


/*
 * This structure describes a simple API-lookup -test.
 *
 * We provide an input and ensure that the output result matches.
 */
typedef struct pwn_test_case
{
    /*
     * Input to the test-case.
     */
    char *input;

    /*
     * Expected result
     */
    int result;

} pwn_test_case;


/*
 * Test that our SHA1-implementation returns somewhat reasonable
 * results.
 */
void test_sha1()
{

    /*
     * Our test-cases - contain the expected input and output
     */
    sha1_test_case input[]  =
    {
        { "steve", "9ce5770b3bb4b2a1d59be2d97e34379cd192299f"},
        { "ssh.pass", "f9ecf6396e3b442df3dae72b81fec784d2b2900d" },
        { "ssh.pasS", "276ed889f7d9a00e24db1c07579f5b78f19ba204"},
        { "x", "11f6ad8ec52a2984abaafd7c3b516503785c2072"},
        { "xx", "dd7b7b74ea160e049dd128478e074ce47254bde8"},
        { "password", "5baa61e4c9b93f3f0682250b6cf8331b7ee68fd8"},
        { "secret", "e5e9fa1ba31ecd1ae84f75caaa474f3a663f05f4" },
        { "25121974", "137babe83f739e3b71211f422fe4c6850f322279"},
    };

    /*
     * For each case.
     */
    size_t cases = sizeof(input) / sizeof(input[0]);

    for (size_t i = 0; i < cases; i++)
    {

        /*
         * Calculate the SHA1-hash of the input, and ensure
         * it matches the expected value.
         */
        unsigned char hash[20];
        SHA1_CTX ctx;
        SHA1Init(&ctx);
        SHA1Update(&ctx, (unsigned char*)input[i].input, strlen(input[i].input));
        SHA1Final(hash, &ctx);

        /*
         * Convert the output to a readable-value.
         */
        char result[41] = {'\0'};

        for (int i = 0; i < 20; i++)
            sprintf(result, "%s%02x", result, hash[i]);


        if (strcmp(input[i].output, result) != 0)
        {
            printf("%zu - FAIL: Test input '%s' gave hash '%s' not '%s'\n",
                   i + 1, input[i].input, result, input[i].output);
            exit(1);
        }
        else
        {
            printf("%zu - OK: Test input '%s' gave expected hash '%s'\n",
                   i + 1, input[i].input, input[i].output);
        }
    }
}


/*
 * Test that the lookup against the remote API looks somewhat sane.
 *
 */
void test_pwn_lookup()
{
    /*
     * Plain-text passwords to lookup, and expected result.
     */
    pwn_test_case input[]  =
    {

        /* Listed in the DB*/
        { "steve", 1},
        { "secret", 1},
        { "secure", 1},
        { "CorrectHorseBatteryStaple", 1},

        /* Not listed in the DB */
        { "fodspfsdpfksdlfdfjlsdfjldfj", 0 },
        { "fdkslf930290kqldsdsfs", 0 },
        { "290809lkfddks,lfdfsdfdsfdsf-_FD-f0s-f09d-0f9sdf0-9q3q12", 0 },
    };

    /*
     * For each case.
     */
    size_t cases = sizeof(input) / sizeof(input[0]);

    for (size_t i = 0; i < cases; i++)
    {

        /*
         * Calculate the SHA1-hash of the input.
         */
        unsigned char hash[20];
        SHA1_CTX ctx;
        SHA1Init(&ctx);
        SHA1Update(&ctx, (unsigned char*)input[i].input, strlen(input[i].input));
        SHA1Final(hash, &ctx);

        /*
         * Convert the output to a readable-value.
         */
        char result[41] = {'\0'};

        for (int i = 0; i < 20; i++)
            sprintf(result, "%s%02x", result, hash[i]);


        /*
         * Now lookup that hash.
         */
        int found = was_leaked(result);

        if (found != input[i].result)
        {
            printf("%zu - FAIL: Test input '%s' gave '%d' not '%d'\n",
                   i + 1, input[i].input, found, input[i].result);
            exit(1);
        }
        else
        {
            printf("%zu - OK: Test input '%s' gave expected result.\n",
                   i + 1, input[i].input);
        }
    }
}



/*
 * Entry-Point.
 */
int main(int argc, char *argv[])
{

    /*
     * Test hash.
     */
    test_sha1();

    /*
     * Test Pwnage.
     */
    test_pwn_lookup();


    exit(0);
}


pwn_chk.c
/*
 * pwn_chk.c - Test an SHA1 hash against the HaveIBeenPwnd API.
 *
 * Steve
 *
 */


#define _XOPEN_SOURCE 700 /* For getline() */

#include <stdio.h>
#include <stdlib.h>
#include <curl/curl.h>
#include <string.h>
#include <syslog.h>
#include <ctype.h>

/*
 * This callback-function is invoked when fresh data has been downloaded
 * via libcurl.
 */
size_t write_data(void *ptr, size_t size, size_t nmemb, FILE *stream)
{
    size_t written = fwrite(ptr, size, nmemb, stream);
    return written;
}

/*
 * Test to see if the given SHA1-hash is known to the HaveIBeenPwnd site.
 *
 * Return value:
 *
 * <0:  Error performing the test.
 *  0:  No leak.
 * >0:  Leaked.
 *
 */
int was_leaked(char *hash)
{

    /*
     * Sanity-check that our input is valid.
     */
    if (hash == NULL  || strlen(hash) != 40)
    {
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: Invalid input for was_leaked(%s).", hash);
        closelog();
        return -1;
    }

    /*
     * We're going to make a request via Curl, so we need a curl
     * object/structure for interfacing with that library.
     */
    CURL *curl;
    CURLcode res;

    /*
     * For the HaveIBeenPwnd API we need to make a request with the
     * first five characters of the SHA1-hash, and then look for the
     * rest of the hash in the response.
     *
     * Extract the first five characters into one buffer, and the rest
     * into another.
     */
    char start[6] = {'\0'};
    strncpy(start, hash, 5);
    char rest[40] = {'\0'};
    strncpy(rest,  hash + 5, sizeof(rest) - 1);

    /*
     * Our hash-prefix should be upper-cased.
     */
    for (unsigned int i = 0; i < strlen(start); i++)
    {
        start[i] = toupper(start[i]);
    }

    /*
     * As should our suffix.
     */
    for (unsigned int i = 0; i < strlen(rest); i++)
    {
        rest[i] = toupper(rest[i]);
    }

    /*
     * The URL we're fetching.
     */
    char url[100] = {'\0'};
    snprintf(url, sizeof(url) - 1,
             "https://api.pwnedpasswords.com/range/%s", start);

    /*
     * Create a temporary file to hold the response.
     */
    FILE * fd = tmpfile();

    /*
     * Initialize curl.
     */
    curl = curl_easy_init();

    if (!curl)
    {
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: Failed to setup curl.");
        closelog();
        fclose(fd);
        return -1;
    }

    /*
     * Setup the request.
     */
    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fd);

    /*
     * Perform the request - this is blocking.
     */
    long res_code = 0;
    res = curl_easy_perform(curl);

    /*
     * Did we receive an error from the remote call?
     */
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &res_code);

    if (!(res_code == 200) && res != CURLE_ABORTED_BY_CALLBACK)
    {

        /*
         * Log the return-code.
         */
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: HTTP Response code was not 200: %ld", res_code);
        closelog();

        curl_easy_cleanup(curl);
        fclose(fd);
        return -1;
    }

    /*
     * Cleanup the curl request.
     */
    curl_easy_cleanup(curl);

    /*
     * At this point we should have a temporary-file containing the result
     * of the request which will be a series of lines of the form:
     *
     *  HASH:COUNT
     *
     * "HASH" is the SHA1-hash (minus the first five characters).
     *
     * "COUNT" is the number of times the hash was leaked (we're going to
     * ignore that, we just care if the password-hash is present at all).
     *
     */

    /*
     * Rewind the file-handle and scan the contents, line-by-line.
     */
    rewind(fd);
    char * line = NULL;
    size_t len = 0;
    ssize_t read;
    int found = 0;

    while ((read = getline(&line, &len, fd)) != -1)
    {
        if (strncmp(rest, line, strlen(rest)) == 0)
        {
            openlog("pam_pwnd", 0, 0);
            syslog(LOG_ERR, "pam_pwnd: Password hash is known-leaked: %s", hash);
            closelog();

            found = 1;
        }
    }

    /*
     * Close the file-handle, which should trigger an unlink too.
     */
    fclose(fd);
    return found;

}


pam_pwnd.c
/*
 * pam_pwnd.c - Test password against the Have I Been Pwnd API.
 *
 * Steve
 *
 */

#define PAM_SM_AUTH

#include <security/pam_modules.h>
#include <security/pam_ext.h>
#include <security/pam_appl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h>
#include <syslog.h>

#include "sha1.h"

/*
 * The lookup of the password is handled via this external function.
 */
extern int was_leaked(char *hash);


/*
 * This function is called to handle the authentication.
 *
 * We first of all retrieve the username, then the password, and after that
 * we run the lookup test.
 */
PAM_EXTERN int pam_sm_authenticate(pam_handle_t * pamh, int flags, int argc, const
                                   char  ** argv)
{
    const char * userName = NULL;
    const char * userPasswd = NULL;

    /*
     * Get the username.
     */
    if (pam_get_user(pamh, &userName, NULL) != PAM_SUCCESS)
    {
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: cannot determine user name.");
        closelog();
        return PAM_USER_UNKNOWN;
    }

    /*
     * Get the password.
     */
    if (pam_get_authtok(pamh, PAM_AUTHTOK, (const char **)&userPasswd, NULL) !=
            PAM_SUCCESS)
    {
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: error getting user password.");
        closelog();
        return PAM_AUTH_ERR;
    }

    /*
     * Sanity-Check
     */
    if (userName == NULL || userPasswd == NULL)
    {
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: cowardly aborting due to null pointer(s).");
        closelog();
        return PAM_AUTHINFO_UNAVAIL;

    }

    /*
     * (SHA1) Hash the password.
     */
    unsigned char hash[20];
    SHA1_CTX ctx;
    SHA1Init(&ctx);
    SHA1Update(&ctx, (unsigned char*)userPasswd, strlen(userPasswd));
    SHA1Final(hash, &ctx);

    /*
     * Convert to a hex-string.
     */
    char buf[41] = {'\0'};

    /*
     * NOTE: We upper-case the string here, but we also repeat that
     * later on.  Just because.
     */
    for (int i = 0; i < 20; i++)
        sprintf(buf, "%s%02X", buf, hash[i]);


    /*
     * Lookup the hash in the remote API.
     */
    int ret = was_leaked(buf);

    /*
     * Handle the result.
     */
    if (ret < 0)
    {

        /*
         * If the return value is <0 that means that something failed.
         *
         * We could deny the login here, but it is better to fail open.
         */
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: couldn't perform hash-test for %s properly.", userName);
        closelog();
        return PAM_SUCCESS;

    }
    else if (ret == 0)
    {

        /*
         * No result :)
         */
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: hash for %s was not listed in remote database (good).", userName);
        closelog();
        return PAM_SUCCESS;
    }
    else
    {

        /*
         * l33t hax0rs have already stolen your password.
         *
         */
        openlog("pam_pwnd", 0, 0);
        syslog(LOG_ERR, "pam_pwnd: hash for %s is included in remote database (bad)!", userName);
        closelog();
    }

    return PAM_AUTHINFO_UNAVAIL;

}


/*
 * Not sure why this is required, if it even is!
 */
PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char
                              *argv[])
{
    return (PAM_SUCCESS);
}


sha1.c
/*
SHA-1 in C
By Steve Reid <steve@edmweb.com>
100% Public Domain

Test Vectors (from FIPS PUB 180-1)
"abc"
  A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
  84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
A million repetitions of "a"
  34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
*/

/* #define LITTLE_ENDIAN * This should be #define'd already, if true. */
/* #define SHA1HANDSOFF * Copies data before messing with it. */

#define SHA1HANDSOFF

#include <stdio.h>
#include <string.h>

/* for uint32_t */
#include <stdint.h>

#include "sha1.h"


#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
#if BYTE_ORDER == LITTLE_ENDIAN
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&0xFF00FF00) \
    |(rol(block->l[i],8)&0x00FF00FF))
#elif BYTE_ORDER == BIG_ENDIAN
#define blk0(i) block->l[i]
#else
#error "Endianness not defined!"
#endif
#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
    ^block->l[(i+2)&15]^block->l[i&15],1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);


/* Hash a single 512-bit block. This is the core of the algorithm. */

void SHA1Transform(
    uint32_t state[5],
    const unsigned char buffer[64]
)
{
    uint32_t a, b, c, d, e;

    typedef union
    {
        unsigned char c[64];
        uint32_t l[16];
    } CHAR64LONG16;

#ifdef SHA1HANDSOFF
    CHAR64LONG16 block[1];      /* use array to appear as a pointer */

    memcpy(block, buffer, 64);
#else
    /* The following had better never be used because it causes the
     * pointer-to-const buffer to be cast into a pointer to non-const.
     * And the result is written through.  I threw a "const" in, hoping
     * this will cause a diagnostic.
     */
    CHAR64LONG16 *block = (const CHAR64LONG16 *) buffer;
#endif
    /* Copy context->state[] to working vars */
    a = state[0];
    b = state[1];
    c = state[2];
    d = state[3];
    e = state[4];
    /* 4 rounds of 20 operations each. Loop unrolled. */
    R0(a, b, c, d, e, 0);
    R0(e, a, b, c, d, 1);
    R0(d, e, a, b, c, 2);
    R0(c, d, e, a, b, 3);
    R0(b, c, d, e, a, 4);
    R0(a, b, c, d, e, 5);
    R0(e, a, b, c, d, 6);
    R0(d, e, a, b, c, 7);
    R0(c, d, e, a, b, 8);
    R0(b, c, d, e, a, 9);
    R0(a, b, c, d, e, 10);
    R0(e, a, b, c, d, 11);
    R0(d, e, a, b, c, 12);
    R0(c, d, e, a, b, 13);
    R0(b, c, d, e, a, 14);
    R0(a, b, c, d, e, 15);
    R1(e, a, b, c, d, 16);
    R1(d, e, a, b, c, 17);
    R1(c, d, e, a, b, 18);
    R1(b, c, d, e, a, 19);
    R2(a, b, c, d, e, 20);
    R2(e, a, b, c, d, 21);
    R2(d, e, a, b, c, 22);
    R2(c, d, e, a, b, 23);
    R2(b, c, d, e, a, 24);
    R2(a, b, c, d, e, 25);
    R2(e, a, b, c, d, 26);
    R2(d, e, a, b, c, 27);
    R2(c, d, e, a, b, 28);
    R2(b, c, d, e, a, 29);
    R2(a, b, c, d, e, 30);
    R2(e, a, b, c, d, 31);
    R2(d, e, a, b, c, 32);
    R2(c, d, e, a, b, 33);
    R2(b, c, d, e, a, 34);
    R2(a, b, c, d, e, 35);
    R2(e, a, b, c, d, 36);
    R2(d, e, a, b, c, 37);
    R2(c, d, e, a, b, 38);
    R2(b, c, d, e, a, 39);
    R3(a, b, c, d, e, 40);
    R3(e, a, b, c, d, 41);
    R3(d, e, a, b, c, 42);
    R3(c, d, e, a, b, 43);
    R3(b, c, d, e, a, 44);
    R3(a, b, c, d, e, 45);
    R3(e, a, b, c, d, 46);
    R3(d, e, a, b, c, 47);
    R3(c, d, e, a, b, 48);
    R3(b, c, d, e, a, 49);
    R3(a, b, c, d, e, 50);
    R3(e, a, b, c, d, 51);
    R3(d, e, a, b, c, 52);
    R3(c, d, e, a, b, 53);
    R3(b, c, d, e, a, 54);
    R3(a, b, c, d, e, 55);
    R3(e, a, b, c, d, 56);
    R3(d, e, a, b, c, 57);
    R3(c, d, e, a, b, 58);
    R3(b, c, d, e, a, 59);
    R4(a, b, c, d, e, 60);
    R4(e, a, b, c, d, 61);
    R4(d, e, a, b, c, 62);
    R4(c, d, e, a, b, 63);
    R4(b, c, d, e, a, 64);
    R4(a, b, c, d, e, 65);
    R4(e, a, b, c, d, 66);
    R4(d, e, a, b, c, 67);
    R4(c, d, e, a, b, 68);
    R4(b, c, d, e, a, 69);
    R4(a, b, c, d, e, 70);
    R4(e, a, b, c, d, 71);
    R4(d, e, a, b, c, 72);
    R4(c, d, e, a, b, 73);
    R4(b, c, d, e, a, 74);
    R4(a, b, c, d, e, 75);
    R4(e, a, b, c, d, 76);
    R4(d, e, a, b, c, 77);
    R4(c, d, e, a, b, 78);
    R4(b, c, d, e, a, 79);
    /* Add the working vars back into context.state[] */
    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;
    state[4] += e;
    /* Wipe variables */
    a = b = c = d = e = 0;
#ifdef SHA1HANDSOFF
    memset(block, '\0', sizeof(block));
#endif
}


/* SHA1Init - Initialize new context */

void SHA1Init(
    SHA1_CTX * context
)
{
    /* SHA1 initialization constants */
    context->state[0] = 0x67452301;
    context->state[1] = 0xEFCDAB89;
    context->state[2] = 0x98BADCFE;
    context->state[3] = 0x10325476;
    context->state[4] = 0xC3D2E1F0;
    context->count[0] = context->count[1] = 0;
}


/* Run your data through this. */

void SHA1Update(
    SHA1_CTX * context,
    const unsigned char *data,
    uint32_t len
)
{
    uint32_t i;

    uint32_t j;

    j = context->count[0];

    if ((context->count[0] += len << 3) < j)
        context->count[1]++;

    context->count[1] += (len >> 29);
    j = (j >> 3) & 63;

    if ((j + len) > 63)
    {
        memcpy(&context->buffer[j], data, (i = 64 - j));
        SHA1Transform(context->state, context->buffer);

        for (; i + 63 < len; i += 64)
        {
            SHA1Transform(context->state, &data[i]);
        }

        j = 0;
    }
    else
        i = 0;

    memcpy(&context->buffer[j], &data[i], len - i);
}


/* Add padding and return the message digest. */

void SHA1Final(
    unsigned char digest[20],
    SHA1_CTX * context
)
{
    unsigned i;

    unsigned char finalcount[8];

    unsigned char c;

#if 0    /* untested "improvement" by DHR */
    /* Convert context->count to a sequence of bytes
     * in finalcount.  Second element first, but
     * big-endian order within element.
     * But we do it all backwards.
     */
    unsigned char *fcp = &finalcount[8];

    for (i = 0; i < 2; i++)
    {
        uint32_t t = context->count[i];

        int j;

        for (j = 0; j < 4; t >>= 8, j++)
            * --fcp = (unsigned char) t
        }

#else

    for (i = 0; i < 8; i++)
    {
        finalcount[i] = (unsigned char)((context->count[(i >= 4 ? 0 : 1)] >> ((3 - (i & 3)) * 8)) & 255);       /* Endian independent */
    }

#endif
    c = 0200;
    SHA1Update(context, &c, 1);

    while ((context->count[0] & 504) != 448)
    {
        c = 0000;
        SHA1Update(context, &c, 1);
    }

    SHA1Update(context, finalcount, 8); /* Should cause a SHA1Transform() */

    for (i = 0; i < 20; i++)
    {
        digest[i] = (unsigned char)
                    ((context->state[i >> 2] >> ((3 - (i & 3)) * 8)) & 255);
    }

    /* Wipe variables */
    memset(context, '\0', sizeof(*context));
    memset(&finalcount, '\0', sizeof(finalcount));
}

void SHA1(
    char *hash_out,
    const char *str,
    int len)
{
    SHA1_CTX ctx;
    unsigned int ii;

    SHA1Init(&ctx);

    for (ii = 0; ii < len; ii += 1)
        SHA1Update(&ctx, (const unsigned char*)str + ii, 1);

    SHA1Final((unsigned char *)hash_out, &ctx);
    hash_out[20] = '\0';
}

